File: Net7 - Copy/040-059/056 VB Pentamino/Piece.vb
' 2006-10-01    PV  VS2005
' 2021-09-18    PV  VS2022, Net6

Public Class Piece
    Public hNumPiece As Short   ' N° dans le jeu Katamino
    Public cPiece As Char   ' Lettre
    Public c As Carre55()    ' 8 transformations maxi
    Public iNbt As Integer   ' Nb de transformations

    '''''''''''''''''''''''''''''''''''
    ' Construction/Destruction
    '''''''''''''''''''''''''''''''''''

    Public Sub New(hNP As Short, cP As Char,
i00 As Integer, i01 As Integer, i02 As Integer, i03 As Integer, i04 As Integer,
i10 As Integer, i11 As Integer, i12 As Integer, i13 As Integer, i14 As Integer,
i20 As Integer, i21 As Integer, i22 As Integer, i23 As Integer, i24 As Integer)
        ReDim c(8 - 1)
        hNumPiece = hNP
        cPiece = cP
        c(0) = New Carre55(i00, i01, i02, i03, i04, i10, i11, i12, i13, i14, i20, i21, i22, i23, i24)
        iNbt = 1
        If i00 + i01 + i02 + i03 + i04 + i10 + i11 + i12 + i13 + i14 + i20 + i21 + i22 + i23 + i24 <> 5 Then
            WriteLine("Définition de la pièce {0} incorrecte", hNP)
        End If

        Dim i, j As Integer
        For i = 1 To 7
            Dim ct As Carre55 = c(0).Transformation(i)
            Dim bDejaVu As Boolean = False

            For j = 0 To iNbt - 1
                If Carre55.Egalite(c(j), ct) Then
                    bDejaVu = True
                    Exit For
                End If
            Next
            If Not bDejaVu Then
                c(iNbt) = ct
                iNbt += 1
            End If
        Next
    End Sub

    ' Traces
    Public Sub Dessin()
        WriteLine("Pièce {0} {1} iNbt={2}", hNumPiece, cPiece, iNbt)
    End Sub

End Class

File: Net8/040-059/056 CS Pentamino/Piece.cs
// Piece.cs
// Représente un pentamino sur une grille 3x5
//
// 2006-10-01   PV		VS 2005
// 2021-09-18	PV		VS2022, Net6
// 2023-01-10	PV		Net7
// 2023-11-18	PV		Net8 C#12

using static System.Console;

internal class Piece
{
    public short hNumPiece;		// N° dans le jeu Katamino
    public char cPiece;			// Lettre
    public Carre55[] c;			// 8 transformations maxi (rotations, symétries)
    public int iNbt;			// Nb de transformations

    public Piece(short hNP, char cP,
                  int i00, int i01, int i02, int i03, int i04,
                  int i10, int i11, int i12, int i13, int i14,
                  int i20, int i21, int i22, int i23, int i24)
    {
        if (i00 + i01 + i02 + i03 + i04 +
            i10 + i11 + i12 + i13 + i14 +
            i20 + i21 + i22 + i23 + i24 != 5)
            WriteLine("Définition de la pièce {0} incorrecte", hNP);

        c = new Carre55[8];
        hNumPiece = hNP;
        cPiece = cP;
        c[0] = new Carre55(i00, i01, i02, i03, i04,
                           i10, i11, i12, i13, i14,
                           i20, i21, i22, i23, i24);
        iNbt = 1;

        for (var i = 1; i < 8; i++)
        {
            var ct = c[0].Transformation(i);
            var bDejaVu = false;

            for (var j = 0; j < iNbt; j++)
                if (Carre55.Egalite(c[j], ct))
                {
                    bDejaVu = true;
                    break;
                }
            if (!bDejaVu)
                c[iNbt++] = ct;
        }
    }

    // For dev/debug traces
    public void Dessin() => WriteLine("Pièce {0} {1} iNbt={2}", hNumPiece, cPiece, iNbt);
}
